### PR TITLE
Vendor gomod dependencies

### PR DESCRIPTION
**Pull request by nirzari**
A data migration is added for adding gomod-vendor flag. When the flag is set, it vendors the dependencies. Another cachito worker config is added to disable/enable cachito strict mode. If the strict mode is enabled and vendor directory exists, cachito requests are failed.

### CODE REVIEW

**Comment by nirzari** [comment]
@mprahl @lcarva Please review!!

**Comment by lcarva** [comment]
@nirzari, could you please address the unit test failures?

File: README.md [hunk 1]
```diff
@@ -93,6 +93,31 @@ For example, on Fedora:
 
 - to add more python dependencies: add to `requirements.txt` and `requirements-workers.txt`
 
+## Database Migrations
```
Comment (lcarva) [inline README.md:96]: It would be best if this was done as a separate commit.

**Comment by nirzari** [comment]
@lcarva Already did! It was due to name change of modules from golang to gomod.

**Comment by mprahl** [comment]
@nirzari could you please document the new flag and explain what it does? A new section called `Flags` would be nice so that additional flags can also be added there.

File: README.md [hunk 2]
```diff
@@ -179,6 +179,8 @@ Custom configuration for the Celery workers are listed below:
   script. This defaults to `1`.
 * `cachito_sources_dir` - the directory for long-term storage of app source archives. This
   configuration is required, and the directory must already exist and be writeable.
+* `cachito_strict_gomod_vendor` - the bool to disable/enable the strict vendor mode. This defaults
```
Comment (mprahl) [inline README.md:182]: Could you elaborate a little more here please?

File: README.md [hunk 3]
```diff
@@ -210,6 +212,11 @@ a web server that supplies the `REMOTE_USER` environment variable when the user
 properly authenticated. A common deployment option is using httpd (Apache web server)
 with the `mod_auth_gssapi` module.
 
+## Flags
+
+* `gomod-vendor` - the flag to indicate the vendoring requirement for gomod dependencies. If present in the
+  cachito request, dependencies are vendored.
```
Comment (mprahl) [inline README.md:218]:
```suggestion
  Cachito request, Cachito will run `go mod vendor` instead of `go mod download` to gather dependencies.
```

File: cachito/workers/pkg_managers/gomod.py [hunk 4]
```diff
@@ -154,19 +165,22 @@ def resolve_gomod(app_source_path, request, dep_replacements=None):
         )
         module = {"name": module_name, "type": "gomod", "version": module_version}
 
-        bundle_dir = RequestBundleDir(request["id"])
+        if "gomod-vendor" not in request.get("flags", {}):
```
Comment (mprahl) [inline cachito/workers/pkg_managers/gomod.py:168]: Same as above
```suggestion
        if "gomod-vendor" not in request.get("flags", []):
```

File: cachito/workers/pkg_managers/gomod.py [hunk 5]
```diff
@@ -154,19 +165,22 @@ def resolve_gomod(app_source_path, request, dep_replacements=None):
...
-            tmp_download_cache_dir,
-            bundle_dir.gomod_download_dir,
-        )
-        shutil.copytree(tmp_download_cache_dir, str(bundle_dir.gomod_download_dir))
+            log.debug(
+                "Adding dependencies from %s to %s",
+                tmp_download_cache_dir,
+                bundle_dir.gomod_download_dir,
+            )
+            shutil.copytree(tmp_download_cache_dir, str(bundle_dir.gomod_download_dir))
```
Comment (mprahl) [inline cachito/workers/pkg_managers/gomod.py:183]: When the flag is provided, could you please create an empty directory at `bundle_dir.gomod_download_dir`? This would make sure that any consumer of Cachito that relies on that directory being present in the bundle would still work.

File: tests/test_workers/test_pkg_managers/test_gomod.py [hunk 6]
```diff
@@ -124,6 +125,122 @@ def test_resolve_gomod(
...
+@mock.patch("os.path.isdir")
+def test_resolve_gomod_strict_mode_raise_error(
+    mock_isdir, mock_gwc, mock_run, mock_temp_dir, tmpdir
+):
+    mock_isdir.return_value = True
+    # Mock the get_worker_config
+    celery_app = celery.Celery()
+    celery_app.cachito_strict_gomod_vendor = True
+    celery_app.cachito_athens_url = "http://athens:3000"
+    mock_gwc.return_value = celery_app
```
Comment (mprahl) [inline tests/test_workers/test_pkg_managers/test_gomod.py:226]: You are setting these attributes on the Celery object and not its configuration. How about just using mock objects instead?
```suggestion
    mock_config = mock.Mock()
    mock_config.cachito_strict_gomod_vendor = True
    mock_config.cachito_athens_url = "http://athens:3000"
    mock_gwc.return_value = mock_config
```

File: tests/test_workers/test_pkg_managers/test_gomod.py [hunk 7]
```diff
@@ -124,6 +125,122 @@ def test_resolve_gomod(
...
+    go_list_error_pkg,
+    expected_replace,
+    tmpdir,
+    sample_deps,
+    sample_deps_replace,
+    sample_deps_replace_new_name,
+    sample_package,
+):
+    mock_cmd_output = _generate_mock_cmd_output(go_list_error_pkg)
+    # mock_gwc.return_value.cachito_strict_gomod_vendor = True
```
Comment (mprahl) [inline tests/test_workers/test_pkg_managers/test_gomod.py:166]:
```suggestion
```

File: tests/test_workers/test_pkg_managers/test_gomod.py [hunk 8]
```diff
@@ -124,6 +125,122 @@ def test_resolve_gomod(
...
+    mock_cmd_output = _generate_mock_cmd_output(go_list_error_pkg)
+    # Mock the tempfile.TemporaryDirectory context manager
+    mock_temp_dir.return_value.__enter__.return_value = str(tmpdir)
+
+    # Mock the "subprocess.run" calls
+    run_side_effects = []
+    if dep_replacement:
```
Comment (mprahl) [inline tests/test_workers/test_pkg_managers/test_gomod.py:173]: Could you remove the `dep_replacement` part of this test since this is well covered in other tests and your code doesn't change that feature? I like to keep tests slim and concise to make future refactoring easier.

File: tests/test_workers/test_pkg_managers/test_gomod.py [hunk 9]
```diff
@@ -124,6 +125,122 @@ def test_resolve_gomod(
...
+            "mod",
+            "edit",
+            "-replace",
+            expected_replace,
+        )
+        if dep_replacement:
+            assert mock_run.call_args_list[2][0][0] == ("go", "mod", "tidy")
+
+    assert module == sample_package
+    assert resolved_deps == expected_deps
```
Comment (mprahl) [inline tests/test_workers/test_pkg_managers/test_gomod.py:211]: Could you test that the `go mod vendor` was called and not `go mod download`?

File: cachito/web/migrations/versions/b46cf36806d7_add_gomod_vendor_flag.py [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+    sa.Column("active", sa.Boolean(), nullable=False, default=True),
+)
+
+
+def upgrade():
+    connection = op.get_bind()
+    # Check if flag exists and is set to false
+    res = connection.execute(
+        flag_table.select().where(
+            flag_table.c.name == "gomod-vendor" and flag_table.c.active == "false"
```
[the remark this answers is not in the document]
Comment (nirzari) [reply cachito/web/migrations/versions/b46cf36806d7_add_gomod_vendor_flag.py]: Oops, updated the code to do the following,
Scenario 1) gomod-vendor exists,
The code will update its value to be `True`
Scenario 2) gomod-vendor doesn't exist,
The code will insert an entry for it and set its value to `True`.

About the second assumption of flag not existing, let's pull lui in and discuss in the next comment!

File: cachito/web/migrations/versions/b46cf36806d7_add_gomod_vendor_flag.py [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+        connection.execute(flag_table.insert().values(name="gomod-vendor", active=True))
+    else:
+        connection.execute(
+            flag_table.update().where(flag_table.c.name == "gomod-vendor").values(active=True)
+        )
+
+
+def downgrade():
+    connection = op.get_bind()
+    connection.execute(flag_table.update().values(name="gomod-vendor", active=False))
```
[the remark this answers is not in the document]
Comment (nirzari) [reply cachito/web/migrations/versions/b46cf36806d7_add_gomod_vendor_flag.py:45]: I had delete and create originally as part of `downgrade` and `upgrade`  but then @lcarva and I discussed this and he suggested to disable-enable the flag instead. It made sense to me at that time but I can't seem to recall the discussion right now. Can you explain please Lui why did we agree on this? (Something to do with the upcoming change of making table name column within flag table as primary key?)

File: cachito/web/migrations/versions/b46cf36806d7_add_gomod_vendor_flag.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (nirzari) [reply cachito/web/migrations/versions/b46cf36806d7_add_gomod_vendor_flag.py]: Resolving as outdated!

File: tests/test_workers/test_pkg_managers/test_gomod.py (hunk 8 above)
Comment (nirzari) [reply tests/test_workers/test_pkg_managers/test_gomod.py:173]: Alright! I was also debating about that but ended up choosing the other way...

File: cachito/workers/pkg_managers/gomod.py [hunk 12]
```diff
@@ -80,9 +80,20 @@ def resolve_gomod(app_source_path, request, dep_replacements=None):
             run_gomod_cmd(
                 ("go", "mod", "edit", "-replace", f"{name}={new_name}@{version}"), run_params
             )
-
-        log.info("Downloading the gomod dependencies")
-        run_gomod_cmd(("go", "mod", "download"), run_params)
+        # Vendor dependencies if the gomod-vendor flag is set
+        if "gomod-vendor" in request.get("flags", {}):
```
[the remark this answers is not in the document]
Comment (nirzari) [reply cachito/workers/pkg_managers/gomod.py:84]: Oops!! Good catch!

File: cachito/workers/pkg_managers/gomod.py [hunk 13]
```diff
@@ -80,9 +80,20 @@ def resolve_gomod(app_source_path, request, dep_replacements=None):
...
+            log.info("Vendoring the gomod dependencies")
+            run_gomod_cmd(("go", "mod", "vendor"), run_params)
+        elif worker_config.cachito_strict_gomod_vendor and os.path.isdir(
+            os.path.join(app_source_path, "vendor")
+        ):
+            raise CachitoError(
+                "Failed to process cachito request as Cachito strict mode for gomod vendor is "
+                "enabled and vendor directory already exists."
```
[the remark this answers is not in the document]
Comment (nirzari) [reply cachito/workers/pkg_managers/gomod.py:92]: Makes more sense! I had altered that msg already 3-4 times as I couldn't make up my mind.

[the remark this answers is not in the document]
Comment (mprahl) [reply cachito/workers/pkg_managers/gomod.py:92]: I just tweaked it again haha

File: cachito/web/migrations/versions/b46cf36806d7_add_gomod_vendor_flag.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (lcarva) [reply cachito/web/migrations/versions/b46cf36806d7_add_gomod_vendor_flag.py:43]: If we delete the flag then a downgrade will cause data loss. That's the only reason.

File: README.md [hunk 14]
```diff
@@ -179,6 +179,10 @@ Custom configuration for the Celery workers are listed below:
   script. This defaults to `1`.
 * `cachito_sources_dir` - the directory for long-term storage of app source archives. This
   configuration is required, and the directory must already exist and be writeable.
+* `cachito_strict_gomod_vendor` - the bool to disable/enable the strict vendor mode. This defaults
+  to `False`. For a repo that has gomod dependencies, if the vendor directory exists and this config
```
Comment (mprahl) [inline README.md:183]:
```suggestion
  to `False`. For a repo that has gomod dependencies, if the `vendor` directory exists and this config
```

File: README.md [hunk 15]
```diff
@@ -179,6 +179,10 @@ Custom configuration for the Celery workers are listed below:
...
+  option is set to `True`, we fail the cachito request. This is added in order to not break the ACM
+  builds.
```
Comment (mprahl) [inline README.md:185]:
```suggestion
  option is set to `True`, Cachito will fail the request.
```

File: cachito/web/migrations/versions/b46cf36806d7_add_gomod_vendor_flag.py [hunk 16]
```diff
@@ -0,0 +1,43 @@
+"""Add the gomod-vendor flag
+
+Revision ID: b46cf36806d7
+Revises: 615c19a1cee1
+Create Date: 2020-05-19 10:33:19.638354
+
```
Comment (mprahl) [inline cachito/web/migrations/versions/b46cf36806d7_add_gomod_vendor_flag.py:6]:
```suggestion
```

File: cachito/web/migrations/versions/b46cf36806d7_add_gomod_vendor_flag.py [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+    sa.MetaData(),
+    sa.Column("id", sa.Integer(), primary_key=True),
+    sa.Column("name", sa.String(), nullable=False),
+    sa.Column("active", sa.Boolean(), nullable=False, default=True),
+)
+
+
+def upgrade():
+    connection = op.get_bind()
+    # Check if flag exists and is set to false
```
Comment (mprahl) [inline cachito/web/migrations/versions/b46cf36806d7_add_gomod_vendor_flag.py:29]:
```suggestion
```

File: cachito/workers/config.py [hunk 18]
```diff
@@ -44,6 +44,7 @@ class Config(object):
     cachito_nexus_timeout = 60
     cachito_nexus_username = "cachito"
     cachito_request_lifetime = 1
+    cachito_strict_gomod_vendor = False
```
Comment (mprahl) [inline cachito/workers/config.py:47]: Could you please rename this to `cachito_gomod_strict_vendor`. I just try to keep configuration options relating to a package manager in the format of `cachito_<package manager>_<configuration option>`.

File: cachito/workers/config.py [hunk 19]
```diff
@@ -66,10 +67,7 @@ class Config(object):
     # Route gomod tasks and npm tasks to separate queues. This is useful if workers are dedicated
     # to specific package managers.
     task_routes = {
-        "cachito.workers.tasks.gomod.*": {
```
Comment (mprahl) [inline cachito/workers/config.py:69]: It's strange that this was not needed before. Could you please move this and the formatting change below to a separate commit?

The reason is that this change is not related to the strict gomod vendor feature. This makes the commit less concise and it also makes it difficult to revert if we ever need to since then we'd need to fix the formatting separately.

File: cachito/workers/pkg_managers/gomod.py [hunk 20]
```diff
@@ -80,9 +80,20 @@ def resolve_gomod(app_source_path, request, dep_replacements=None):
             run_gomod_cmd(
                 ("go", "mod", "edit", "-replace", f"{name}={new_name}@{version}"), run_params
             )
-
-        log.info("Downloading the gomod dependencies")
-        run_gomod_cmd(("go", "mod", "download"), run_params)
+        # Vendor dependencies if the gomod-vendor flag is set
+        if "gomod-vendor" in request.get("flags", []):
```
Comment (mprahl) [inline cachito/workers/pkg_managers/gomod.py:84]: Optional:
```suggestion
        flags = request.get("flags", [])
        if "gomod-vendor" in flags:
```

File: cachito/workers/pkg_managers/gomod.py [hunk 21]
```diff
@@ -156,17 +167,22 @@ def resolve_gomod(app_source_path, request, dep_replacements=None):
 
         bundle_dir = RequestBundleDir(request["id"])
 
-        # Add the gomod cache to the bundle the user will later download
-        tmp_download_cache_dir = os.path.join(temp_dir, RequestBundleDir.go_mod_cache_download_part)
-        if not os.path.exists(tmp_download_cache_dir):
-            os.makedirs(tmp_download_cache_dir, exist_ok=True)
+        if "gomod-vendor" in request.get("flags", []):
```
Comment (mprahl) [inline cachito/workers/pkg_managers/gomod.py:170]: Optional: If you accept the previous suggestion.
```suggestion
        if "gomod-vendor" in flags:
```

File: cachito/workers/pkg_managers/gomod.py [hunk 22]
```diff
@@ -156,17 +167,22 @@ def resolve_gomod(app_source_path, request, dep_replacements=None):
...
+            bundle_dir.gomod_download_dir.mkdir(exist_ok=True, parents=True)
```
Comment (mprahl) [inline cachito/workers/pkg_managers/gomod.py:171]:
```suggestion
            # Create an empty gomod cache in the bundle directory so that any Cachito
            # user does not have to guard against this directory not existing
            bundle_dir.gomod_download_dir.mkdir(exist_ok=True, parents=True)
```

File: README.md [hunk 23]
```diff
@@ -179,6 +179,9 @@ Custom configuration for the Celery workers are listed below:
   script. This defaults to `1`.
 * `cachito_sources_dir` - the directory for long-term storage of app source archives. This
   configuration is required, and the directory must already exist and be writeable.
+* `cachito_gomod_strict_vendor` - the bool to disable/enable the strict vendor mode. This defaults
+  to `False`. For a repo that has gomod dependencies, if the `vendor` directory exists and this config
+  option is set to `True`,Cachito will fail the request.
```
Comment (mprahl) [inline README.md:184]:
```suggestion
  option is set to `True`, Cachito will fail the request.
```

**Comment by mprahl** [comment]
@nirzari looks good! I just left one last minor comment.

@lcarva could you please review this again since you were the original requester?
